applications/services/gui/elements.c

```diff
@@ -605,22 +617,53 @@ void elements_scrollable_text_line(
         // Calculate scroll size
         size_t scroll_size = furi_string_size(line);
         size_t right_width = 0;
-        for(size_t i = scroll_size; i > 0; i--) {
-            right_width += canvas_glyph_width(canvas, furi_string_get_char(line, i));
+        for(size_t i = scroll_size - 1; i > 0; i--) {
+            while((furi_string_get_char(line, i) & 0xC0) == 0x80) {
+                i--;
+            }
+
+            FuriStringUTF8State state = FuriStringUTF8StateStarting;
+            FuriStringUnicodeValue value = 0;
+
+            size_t offset = i;
+            do {
+                furi_string_utf8_decode(furi_string_get_char(line, offset), &state, &value);
+                offset++;
+            } while(state != FuriStringUTF8StateStarting);
+
+            right_width += canvas_glyph_width(canvas, value);
             if(right_width > width) break;
             scroll_size--;
             if(!scroll_size) break;
         }
+
         // Ensure that we have something to scroll
         if(scroll_size) {
             scroll_size += 3;
             scroll = scroll % scroll_size;
-            furi_string_right(line, scroll);
+
+            size_t offset = 0;
+            for(; scroll > 0; offset++) {
+                if((furi_string_get_char(line, offset) & 0xC0) != 0x80) {
+                    scroll--;
+                }
+            }
+
+            // Step back to the end of the previous symbol
+            offset--;
+            furi_string_right(line, offset);
         }
 
         len_px = canvas_string_width(canvas, furi_string_get_cstr(line));
         while(len_px > width) {
-            furi_string_left(line, furi_string_size(line) - 1);
+            size_t offset = 1;
+            size_t len = furi_string_size(line);
+
+            while((furi_string_get_char(line, len - offset) & 0xC0) == 0x80) {
+                offset++;
```

Comment:
Whole pointer arithmetic thing looks soooooo sketchy and unreliable. Clearly something must be done with it before we will be able to merge it.